libethcore/Farm.h

```diff
@@ -83,6 +83,10 @@ class GenericFarm: public GenericFarmFace<PoW>
 	 */
 	bool start(std::string const& _sealer)
 	{
+		// call empty setWork to make sure that all workers have stoppedWorking()
+		for (auto& miner: m_miners)
+			miner->setWork();
+
```

Comment:
The problem is this `WriteGuard`. The deadlock happens because this is held and is waiting for stopWorking to be called to release it. At the same time other threads are also waiting on the `WriteGuard` to be released. So somehow they need to cleared (ensure that stopWorking() has been called for all of them)  before the writeguard is held.

It's easy to reproduce this deadlock. Just try to GPU mine with a synced blockchain on develop. After a while this will occur.